productivity/driver.py

```diff
@@ -106,53 +131,69 @@ def get_tags(self):
         for key, value in to_write.items():
+            start_address = self.tags[key]['address']['start']
             data_type = self.tags[key]['type']
             if isinstance(value, int) and data_type == 'float':
                 to_write[key] = float(value)
             if not isinstance(value, pydoc.locate(data_type)):
                 raise ValueError(f"Expected {key} to be a {data_type}.")
-        addresses = [(tag, self.tags[tag]['address']['start']) for tag in to_write]
-        if any([100000 <= x[1] < 400000 for x in addresses]):
-            raise ValueError("Cannot write to input registers.")
-        addresses.sort(key=lambda a: a[1])
-
-        responses = []
-        while addresses and addresses[0][1] < 65536:
-            address = addresses.pop(0)
-            responses.append(str(await self.write_coil(address[1] - 1,
-                                                       to_write[address[0]])))
-        while addresses and 400000 <= addresses[0][1] < 465536:
-            address = addresses.pop(0)
-            builder = BinaryPayloadBuilder(byteorder=Endian.Big,
-                                           wordorder=Endian.Little)
-            key = address[0]
-            value = to_write[key]
-            data_type = self.tags[key]['type']
-            if data_type == 'float':
-                builder.add_32bit_float(value)
-            elif data_type == 'str':
-                chars = self.tags[key]['length']
-                if len(value) > chars:
-                    raise ValueError(f'{value} is too long for {key}. '
-                                     f'Max: {chars} chars')
-                builder.add_string(value.ljust(chars))
-            elif data_type == 'int':
-                builder.add_16bit_int(value)
-            elif data_type == 'int32':
-                builder.add_32bit_int(value)
+            if 0 <= start_address < 65536:
+                discrete_to_write[key] = value
+            elif 400000 <= start_address < 465536:
+                registers_to_write[key] = value
             else:
-                raise ValueError("Missing data type.")
-            resp = await self.write_registers(address[1] - 400001,
-                                              builder.build(),
-                                              skip_encode=True)
-            responses.append(str(resp[0]))
-        if addresses:
-            raise ValueError("Not all registers spent.")
-        return responses
+                ValueError(f"{key} is not at a writeable address: {start_address}")
+        return discrete_to_write, registers_to_write
+
+    async def _write_register_value(self, key: str,
+                                    value: Union[str, float, int]
+                                    ) -> WriteMultipleRegistersResponse:
+        """Write a single value to the holding registers.
+
+        Currently registers are written one at a time to avoid issues with
+        discontinuous modbus addresses.
```

Comment:
We discussed, and I agree that this is probably okay for all envisioned use cases. Let's just keep this implementation in mind for future control system code reviews.